### PR TITLE
SVPG & ADR

### PR DESCRIPTION
**Pull request by rmenzenbach**
Format

SVPG bugfixes

SVPG docs

Cleanup

Style

ADR hyperparams

Small Changes

Improved reward generator

ADR experiment

MP and hyper parameters

Refactoring

Adr structure

Fixed ADR Adapter

Cleanup

Experiment

Test saving of discriminator

### CODE REVIEW

File: Pyrado/pyrado/algorithms/step_based/svpg.py [hunk 1]
```diff
@@ -66,6 +72,7 @@ def __init__(
         Constructor
 
         :param save_dir: directory to save the snapshots i.e. the results in
+        :type save_dir: pyrado.PathLike
```
Comment (famura) [inline Pyrado/pyrado/algorithms/step_based/svpg.py:75]: please remove type hints in the doc strings

File: Pyrado/pyrado/algorithms/step_based/svpg.py [hunk 2]
```diff
@@ -154,61 +105,50 @@ def zero_grad(self, *args, **kwargs):
...
                 parameters[i].append(params)
                 args[i].append(args_i)
                 kwargs[i].append(kwargs_i)
+            particle.optim.reset()
+            self.logger.add_value("particle", i)
+            avg_ret = self.logger._current_values["avg return"]
+            self.make_snapshot(snapshot_mode, avg_ret, meta_info)
+            self.logger.record_step()
 
         # assert all(len(p) == len(parameters[0]) for p in parameters)
```
Comment (famura) [inline Pyrado/pyrado/algorithms/step_based/svpg.py:141]: please remove if unsued

File: Pyrado/pyrado/algorithms/step_based/svpg.py [hunk 3]
```diff
@@ -281,3 +219,77 @@ def store_particle(self):
...
         self.particle_states[self.current_particle] = self.particle.__getstate__()
         self.particle_policy_states[self.current_particle] = to.clone(self.particle.policy.param_values)
+
+
+# class SVPGHyperparams(TypedDict):
+#    algo: Dict
+#    actor: Dict
+#    vfcn: Dict
+#    critic: Dict
+#    particle: Dict
```
Comment (famura) [inline Pyrado/pyrado/algorithms/step_based/svpg.py:229]: please remove if unused

File: Pyrado/scripts/training/qq-su_adr.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    # Environment
+    env_hparams = dict(dt=1 / 100.0, max_steps=600)
+    env = QQubeSwingUpSim(**env_hparams)
+    env = ActNormWrapper(env)
+
+    ##
+    #
+    # Subroutine
+    #
+    ##
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_adr.py:71]: please shrink to one line

File: Pyrado/scripts/training/qq-su_adr.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        lr_scheduler=lr_scheduler.ExponentialLR,
+        lr_scheduler_hparam=dict(gamma=0.999),
+    )
+    subroutine = PPO(ex_dir, env, policy, critic, **subroutine_hparam)
+
+    ###
+    #
+    # ADR
+    #
+    ###
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_adr.py:113]: please shrink to one line

**Review by famura — CHANGES_REQUESTED** [review]
Very nice! Just a few minor things

**Review by famura — APPROVED** [review]
